doc/internals.rst

```diff
@@ -231,3 +231,200 @@ re-open it directly with Zarr:
+- object reference: the reference of the class that you have implemented.
+
+See https://packaging.python.org/specifications/entry-points/#data-model
+for more information
+
+How to support Lazy Loading
++++++++++++++++++++++++++++
+If you want to make your backend effective with big datasets, then you should support
+the lazy loading.
+Basically, you shall replace the :py:class:`numpy.array` inside the variables with
+a custom class:
+
+.. ipython:: python
+    backend_array = YourBackendArray()
+    data = indexing.LazilyOuterIndexedArray(backend_array)
+    variable = Variable(..., data, ...)
+
+Where ``YourBackendArray``is a class that inherit from
+:py:class:`~xarray.backends.common.BackendArray` and
+:py:class:`~xarray.core.indexing.LazilyOuterIndexedArray` is a
+class of Xarray that wraps an array to make basic and outer indexing lazy.
+
+BackendArray
+^^^^^^^^^^^^
+
+CachingFileManager
+^^^^^^^^^^^^^^^^^^
+
+
+Dask chunking
++++++++++++++
+The backend is not directly involved in `Dask <http://dask.pydata.org/>`__ chunking, since it is managed
+internally by Xarray. However, the backend can define the preferred chunk size
+inside the variable’s encoding ``var.encoding["preferred_chunks"]``.
+The ``preferred_chunks`` may be useful to improve performances with lazy loading.
+``preferred_chunks`` shall be a  dictionary specifying chunk size per dimension
+like ``{“dim1”: 1000, “dim2”: 2000}``  or
+``{“dim1”: [1000, 100], “dim2”: [2000, 2000, 2000]]}``.
+
+The ``preferred_chunks`` is used by Xarray to define the chunk size in some
+special cases:
+
+- If ``chunks`` along a dimension is None or not defined
+- If ``chunks`` is “auto”
+
+In the first case Xarray uses the chunks size specified in
+``preferred_chunks``.
+In the second case Xarray accommodates ideal chunk sizes, preserving if
+possible the "preferred_chunks". The ideal chunk size is computed using
+``dask.core.normalize function``, setting ``previus_chunks = preferred_chunks``.
+
+
+Decoders
+++++++++
+The decoders implement specific operations to transform data from on-disk
+representation to Xarray representation.
+
+A classic example is the “time” variable decoding operation. In NetCDF, the
+elements of the “time” variable are stored as integers, and the unit contains
+an origin (for example: "seconds since 1970-1-1"). In this case, Xarray
+transforms the pair integer-unit in a ``np.datetimes``.
+
+The standard decoders implemented in Xarray are:
+- strings.CharacterArrayCoder()
+- strings.EncodedStringCoder()
+- variables.UnsignedIntegerCoder()
+- variables.CFMaskCoder()
+- variables.CFScaleOffsetCoder()
+- times.CFTimedeltaCoder()
+- times.CFDatetimeCoder()
+
+Some of the transformations can be common to more backends, so before
+implementing a new decoder, be sure Xarray does not already implement that one.
+
+The backends can reuse Xarray’s decoders, either instantiating the decoders
+directly or using the higher-level function
```

Comment:
Here again an example showing the use of one built-in decoder function would be useful.

Reply:
done